### PR TITLE
[WIP] Added api to create evaluation.

### PR DESCRIPTION
**Pull request by lu-ohai**
### Added api to create evaluation.

- Currently waiting for the evaluation job yaml to be completed and this PR runs a dummy job and is for the purpose of end to end demo.

### CODE REVIEW

File: ads/aqua/evaluation.py [hunk 1]
```diff
@@ -85,35 +139,327 @@ class AquaEvaluationApp(AquaApp):
...
-            },
-            "job": {
-                "id": "ocid1.datasciencejob.<OCID>",
-                "name": "test_evaluation",
-                "console_url": "https://cloud.oracle.com/data-science/jobs/ocid1.datasciencejob.<OCID>?region=us-ashburn-1",
-            },
-            "tags": {
-                "aqua_evaluation": "aqua_evaluation",
-            },
+    def create(
```
Comment (mayoor) [inline ads/aqua/evaluation.py:142]: These are too many parameters in a function. We should group them into dataclass to make it readable.

File: ads/aqua/extension/evaluation_handler.py [hunk 2]
```diff
@@ -21,8 +24,51 @@ def get(self, eval_id=""):
 
     @handle_exceptions
     def post(self, *args, **kwargs):
```
Comment (mayoor) [inline ads/aqua/extension/evaluation_handler.py:26]: Is there a way to use dataclass and automatic deserialization provided by pydantic/tornado?

File: ads/aqua/evaluation.py [hunk 3]
```diff
@@ -85,35 +139,327 @@ class AquaEvaluationApp(AquaApp):
...
+                    ocpus=ocpus
+                )
+                .with_block_storage_size(block_storage_size)
+                .with_freeform_tag(
+                    **evaluation_job_freeform_tags
+                )
+            )
+            .with_runtime(
+                # TODO: for the purpose of demo and will revist after
+                PythonRuntime()
```
Comment (mayoor) [inline ads/aqua/evaluation.py:369]: can we have a separate method which will return the runtime. That way when we change the logic, we dont have to touch this method/file. Might also help us with testing with alternate runtimes in future.

File: ads/aqua/utils.py [hunk 4]
```diff
@@ -20,6 +25,7 @@
 README = "README.md"
 DEPLOYMENT_CONFIG = "deployment_config.json"
```
Comment (mayoor) [inline ads/aqua/utils.py:26]: let us add `os.environ.get("AQUA_DEPLOYMENT_CONFIG", "deployment_config.json")`

File: ads/aqua/utils.py [hunk 5]
```diff
@@ -20,6 +25,7 @@
...
 UNKNOWN_JSON_STR = "{}"
+SUPPORTED_FILE_FORMATS = [".jsonl"]
```
Comment (mayoor) [inline ads/aqua/utils.py:28]: we should drop the `.`. We can keep it just jsonl

File: ads/aqua/utils.py [hunk 6]
```diff
@@ -108,3 +114,26 @@ def read_file(file_path: str, **kwargs) -> str:
...
+):
+    expanded_path = os.path.expanduser(src_uri)
+    if not os.path.isfile(expanded_path):
+        raise AquaFileNotFoundError("Invalid input file path. Specify a valid one.")
+    if Path(expanded_path).suffix not in SUPPORTED_FILE_FORMATS:
+        raise AquaValueError(
+            f"Invalid input file. Only {', '.join(SUPPORTED_FILE_FORMATS)} files are supported."
+        )
+    if os.path.getsize(expanded_path) == 0:
+        raise AquaValueError("Empty input file. Specify a non-empty one.")
```
Comment (mayoor) [inline ads/aqua/utils.py:132]:
```suggestion
        raise AquaValueError("Empty input file. Specify a valid file path.")
```

File: ads/aqua/extension/evaluation_handler.py [hunk 7]
```diff
@@ -21,8 +24,51 @@ def get(self, eval_id=""):
...
+                    evaluation_description=input_data.get("evaluation_description"),
+                    project_id=input_data.get("project_id", PROJECT_OCID),
+                    dataset_path=input_data.get("dataset_path"),
+                    report_path=input_data.get("report_path"),
+                    model_parameters=input_data.get("model_parameters"),
+                    shape_name=input_data.get("shape_name"),
+                    memory_in_gbs=input_data.get("memory_in_gbs"),
+                    ocpus=input_data.get("ocpus"),
+                    block_storage_size=input_data.get("block_storage_size"),
+                    compartment_id=input_data.get("compartment_id", COMPARTMENT_OCID),
```
Comment (mayoor) [inline ads/aqua/extension/evaluation_handler.py:61]: We have the default logic while creating the job right? We can just pass `input_data.get("compartment_id")`

File: ads/aqua/evaluation.py (hunk 1 above)
Comment (lu-ohai) [reply ads/aqua/evaluation.py:210]: Thanks, updated to use new data class `CreateAquaEvaluationDetails` which contains all required and optional fields to initialize the creation.

File: ads/aqua/evaluation.py [hunk 8]
```diff
@@ -85,35 +139,327 @@ class AquaEvaluationApp(AquaApp):
...
+        )
+        logger.debug(
+            f"Successfully created evaluation model {evaluation_model.id} for {evaluation_source_id}."
+        )
+
+        if not ObjectStorageDetails.is_oci_path(report_path):
+            raise AquaValueError("Evaluation report path must be an object storage path.")
+
+        evaluation_dataset_path = dataset_path
+        if not ObjectStorageDetails.is_oci_path(evaluation_dataset_path):
```
[the remark this answers is not in the document]
Comment (lu-ohai) [reply ads/aqua/evaluation.py:354]: This validation can't be moved up as it needs the evaluation model id which must be placed after the creation of model.

File: ads/aqua/model.py [hunk 9]
```diff
@@ -134,7 +134,11 @@ def create(
             .with_defined_metadata_list(service_model.defined_metadata_list)
             .with_provenance_metadata(service_model.provenance_metadata)
             # TODO: decide what kwargs will be needed.
-            .create(model_by_reference=True, **kwargs)
+            .create(
+                remove_existing_artifact=False,
```
[the remark this answers is not in the document]
Comment (lu-ohai) [reply ads/aqua/model.py:138]: I'm not sure, could you confirm that `remove_existing_artifact=False` is not needed for model by reference?

[the remark this answers is not in the document]
Comment (VipulMascarenhas) [reply ads/aqua/model.py:138]: yea, we don't need it.

**Review by VipulMascarenhas — APPROVED** [review]

File: ads/aqua/utils.py [hunk 10]
```diff
@@ -109,3 +114,26 @@ def read_file(file_path: str, **kwargs) -> str:
     except Exception as e:
         logger.error(f"Failed to read file {file_path}. {e}")
         return UNKNOWN
+    
+def upload_file_to_os(
```
Comment (mingkang111) [inline ads/aqua/utils.py:118]: Can the `ads.common.utils.copy_file` being used instead?

**Pull request merged**